File: agora/store.py
"""
Agora local message store.

Messages are persisted locally so chat history survives relay outages.
The store is a directory of JSON files, one per message.

Storage locations:
  ~/.agora/rooms/<room_id>/messages/  — message files
  ~/.agora/rooms/<room_id>/state.json — room state (ratchet position, seen IDs)
  ~/.agora/rooms.json                 — room registry
  ~/.agora/identity.json              — agent identity
"""

import json
import os
import time
from pathlib import Path
from typing import Optional


AGORA_DIR = Path.home() / ".agora"


def _ensure(path: Path):
    path.mkdir(parents=True, exist_ok=True)


# ── Identity ────────────────────────────────────────────────────

def get_agent_id() -> str:
    """Get or create a persistent agent identity."""
    identity_file = AGORA_DIR / "identity.json"
    if identity_file.exists():
        data = json.loads(identity_file.read_text())
        return data.get("agent_id", "")

    # Derive from session or generate
    sid = os.environ.get("CLAUDE_CODE_SESSION_ID", "")
    if sid.startswith("cse_"):
        agent_id = sid[4:12]
    elif sid:
        agent_id = sid[:8]
    else:
        import secrets
        agent_id = secrets.token_hex(4)

    _ensure(AGORA_DIR)
    identity_file.write_text(json.dumps({"agent_id": agent_id}))
    return agent_id


# ── Room Registry ───────────────────────────────────────────────

def registry_path() -> Path:
    return AGORA_DIR / "rooms.json"


def load_registry() -> list[dict]:
    """Load the room registry."""
    path = registry_path()
    if path.exists():
        try:
            return json.loads(path.read_text())
        except json.JSONDecodeError:
            pass
    return []


def save_registry(rooms: list[dict]):
    """Save the room registry."""
    _ensure(AGORA_DIR)
    registry_path().write_text(json.dumps(rooms, indent=2) + "\n")


def add_room(room_id: str, secret: str, label: str) -> dict:
    """Add a room to the registry. Returns the room entry."""
    rooms = load_registry()
    for r in rooms:
        if r["room_id"] == room_id:
            return r  # already exists
    entry = {
        "room_id": room_id,
        "secret": secret,
        "label": label,
        "joined_at": int(time.time()),
    }
    rooms.append(entry)
    save_registry(rooms)
    return entry


def find_room(label_or_id: str) -> Optional[dict]:
    """Find a room by label or room_id."""
    for r in load_registry():
        if r["label"] == label_or_id or r["room_id"] == label_or_id:
            return r
    return None


def get_active_room() -> Optional[dict]:
    """Get the currently active room."""
    state_file = AGORA_DIR / "active_room"
    if state_file.exists():
        label = state_file.read_text().strip()
        return find_room(label)
    rooms = load_registry()
    return rooms[0] if rooms else None


def set_active_room(label: str):
    """Set the active room."""
    _ensure(AGORA_DIR)
    (AGORA_DIR / "active_room").write_text(label)


# ── Message Persistence ─────────────────────────────────────────

def _msg_dir(room_id: str) -> Path:
    return AGORA_DIR / "rooms" / room_id / "messages"


def save_message(room_id: str, envelope: dict):
    """Persist a decrypted message envelope."""
    d = _msg_dir(room_id)
    _ensure(d)
    ts = envelope.get("ts", int(time.time()))
    mid = envelope.get("id", "x")
    path = d / f"{ts}_{mid}.json"
    if not path.exists():
        path.write_text(json.dumps(envelope) + "\n")


def load_messages(room_id: str, since_secs: int = 7200) -> list[dict]:
    """Load messages from local store within time window."""
    d = _msg_dir(room_id)
    if not d.exists():
        return []
    cutoff = int(time.time()) - since_secs
    msgs = []
    for f in sorted(d.glob("*.json")):
        try:
            env = json.loads(f.read_text().strip())
            if env.get("ts", 0) >= cutoff:
                msgs.append(env)
        except (json.JSONDecodeError, OSError):
            pass
    return msgs


# ── Seen Message Tracking ───────────────────────────────────────

def _seen_path(room_id: str) -> Path:
    return AGORA_DIR / "rooms" / room_id / "seen.txt"


def load_seen(room_id: str) -> set[str]:
    """Load set of seen message IDs."""
    path = _seen_path(room_id)
    if path.exists():
        return set(path.read_text().strip().split("\n"))
    return set()


def mark_seen(room_id: str, msg_id: str):
    """Mark a message ID as seen."""
    path = _seen_path(room_id)
    _ensure(path.parent)
    seen = load_seen(room_id)
    seen.add(msg_id)
    # Keep only last 1000
    recent = sorted(seen)[-1000:]
    path.write_text("\n".join(recent))


# ── Room State (ratchet position, etc.) ─────────────────────────

def _state_path(room_id: str) -> Path:
    return AGORA_DIR / "rooms" / room_id / "state.json"


def load_room_state(room_id: str) -> dict:
    path = _state_path(room_id)
    if path.exists():
        try:
            return json.loads(path.read_text())
        except json.JSONDecodeError:
            pass
    return {}


def save_room_state(room_id: str, state: dict):
    path = _state_path(room_id)
    _ensure(path.parent)
    path.write_text(json.dumps(state, indent=2))

File: agora-rs/src/store.rs
//! Agora local message store.
//!
//! Persistence at ~/.agora/:
//!   rooms/<room_id>/messages/ — message files
//!   rooms/<room_id>/seen.txt  — seen message IDs
//!   rooms.json                — room registry
//!   identity.json             — agent identity

use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::fs;
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};

fn agora_dir() -> PathBuf {
    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("/tmp"))
        .join(".agora")
}

fn ensure_dir(path: &PathBuf) {
    let _ = fs::create_dir_all(path);
}

fn now() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs()
}

// ── Identity ────────────────────────────────────────────────────

pub fn get_agent_id() -> String {
    let id_file = agora_dir().join("identity.json");
    if let Ok(data) = fs::read_to_string(&id_file) {
        if let Ok(v) = serde_json::from_str::<serde_json::Value>(&data) {
            if let Some(id) = v["agent_id"].as_str() {
                return id.to_string();
            }
        }
    }

    // Derive from env or generate
    let agent_id = if let Ok(sid) = std::env::var("CLAUDE_CODE_SESSION_ID") {
        if sid.starts_with("cse_") {
            sid[4..12.min(sid.len())].to_string()
        } else {
            sid[..8.min(sid.len())].to_string()
        }
    } else {
        let rng = ring::rand::SystemRandom::new();
        let mut buf = [0u8; 4];
        ring::rand::SecureRandom::fill(&rng, &mut buf).expect("RNG failed");
        hex::encode(buf)
    };

    let dir = agora_dir();
    ensure_dir(&dir);
    let json = serde_json::json!({"agent_id": agent_id});
    let _ = fs::write(&id_file, serde_json::to_string_pretty(&json).unwrap());
    agent_id
}

// ── Room Registry ───────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoomEntry {
    pub room_id: String,
    pub secret: String,
    pub label: String,
    pub joined_at: u64,
}

pub fn load_registry() -> Vec<RoomEntry> {
    let path = agora_dir().join("rooms.json");
    if let Ok(data) = fs::read_to_string(&path) {
        serde_json::from_str(&data).unwrap_or_default()
    } else {
        vec![]
    }
}

pub fn save_registry(rooms: &[RoomEntry]) {
    let dir = agora_dir();
    ensure_dir(&dir);
    let data = serde_json::to_string_pretty(rooms).unwrap();
    let _ = fs::write(dir.join("rooms.json"), data);
}

pub fn add_room(room_id: &str, secret: &str, label: &str) -> RoomEntry {
    let mut rooms = load_registry();
    if let Some(existing) = rooms.iter().find(|r| r.room_id == room_id) {
        return existing.clone();
    }
    let entry = RoomEntry {
        room_id: room_id.to_string(),
        secret: secret.to_string(),
        label: label.to_string(),
        joined_at: now(),
    };
    rooms.push(entry.clone());
    save_registry(&rooms);
    entry
}

pub fn find_room(label_or_id: &str) -> Option<RoomEntry> {
    load_registry()
        .into_iter()
        .find(|r| r.label == label_or_id || r.room_id == label_or_id)
}

pub fn get_active_room() -> Option<RoomEntry> {
    let active_file = agora_dir().join("active_room");
    if let Ok(label) = fs::read_to_string(&active_file) {
        let label = label.trim();
        if let Some(room) = find_room(label) {
            return Some(room);
        }
    }
    load_registry().into_iter().next()
}

pub fn set_active_room(label: &str) {
    let dir = agora_dir();
    ensure_dir(&dir);
    let _ = fs::write(dir.join("active_room"), label);
}

// ── Message Persistence ─────────────────────────────────────────

pub fn save_message(room_id: &str, envelope: &serde_json::Value) {
    let dir = agora_dir()
        .join("rooms")
        .join(room_id)
        .join("messages");
    ensure_dir(&dir);
    let ts = envelope["ts"].as_u64().unwrap_or_else(now);
    let mid = envelope["id"].as_str().unwrap_or("x");
    let path = dir.join(format!("{ts}_{mid}.json"));
    if !path.exists() {
        let _ = fs::write(&path, serde_json::to_string(envelope).unwrap());
    }
}

pub fn load_messages(room_id: &str, since_secs: u64) -> Vec<serde_json::Value> {
    let dir = agora_dir()
        .join("rooms")
        .join(room_id)
        .join("messages");
    if !dir.exists() {
        return vec![];
    }
    let cutoff = now().saturating_sub(since_secs);
    let mut msgs = Vec::new();
    if let Ok(entries) = fs::read_dir(&dir) {
        for entry in entries.flatten() {
            if let Ok(data) = fs::read_to_string(entry.path()) {
                if let Ok(env) = serde_json::from_str::<serde_json::Value>(&data) {
                    if env["ts"].as_u64().unwrap_or(0) >= cutoff {
                        msgs.push(env);
                    }
                }
            }
        }
    }
    msgs.sort_by_key(|m| m["ts"].as_u64().unwrap_or(0));
    msgs
}

// ── Seen Tracking ───────────────────────────────────────────────

pub fn load_seen(room_id: &str) -> HashSet<String> {
    let path = agora_dir().join("rooms").join(room_id).join("seen.txt");
    if let Ok(data) = fs::read_to_string(&path) {
        data.lines().map(|s| s.to_string()).collect()
    } else {
        HashSet::new()
    }
}

pub fn mark_seen(room_id: &str, msg_id: &str) {
    let dir = agora_dir().join("rooms").join(room_id);
    ensure_dir(&dir);
    let path = dir.join("seen.txt");
    let mut seen = load_seen(room_id);
    seen.insert(msg_id.to_string());
    let mut ids: Vec<_> = seen.into_iter().collect();
    ids.sort();
    if ids.len() > 1000 {
        ids = ids[ids.len() - 1000..].to_vec();
    }
    let _ = fs::write(&path, ids.join("\n"));
}
